clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs

```diff
@@ -0,0 +1,387 @@
+  { cfFlags :: IORef ClashOpts -> [Flag IO]
+  , cfRender :: ClashOpts -> [String]
+  }
+
+type FlagName = String
+
+--------------------------
+-- Simple flag builders --
+--------------------------
+
+-- | Return two flags covering a single property: @-fclash-foo@ and @-fclash-no-foo@.
+boolFlag :: FlagName -> Lens' ClashOpts Bool -> ClashFlag
+boolFlag nm opt = ClashFlag
+  { cfFlags = \opts -> boolArg nm opts (set opt)
+  , cfRender = \opts -> if isDefault opt opts then [] else renderNoArg nm (opts ^. opt)
+  }
+
+-- | Return a flag setting a 'Word': @-fclash-foo=3@.
+wordFlag :: FlagName -> Lens' ClashOpts Word -> ClashFlag
+wordFlag nm opt = ClashFlag
+  { cfFlags = \ref -> wordSuffixId nm ref (set opt)
+  , cfRender = \opts ->
+      if isDefault opt opts then [] else renderSuffix nm (show (opts ^. opt))
+  }
+
+-- | Return a flag setting a 'Int': @-fclash-foo=3@.
+intFlag :: FlagName -> Lens' ClashOpts Int -> ClashFlag
+intFlag nm opt = ClashFlag
+  { cfFlags = \ref -> intSuffixId nm ref (set opt)
+  , cfRender = \opts ->
+      if isDefault opt opts then [] else renderSuffix nm (show (opts ^. opt))
+  }
+
+-- | Return two flags setting a 'Maybe' 'String': @-fclash-foo wibble@ and
+-- @-fclash-no-foo@.
+maybeStringFlag :: FlagName -> Lens' ClashOpts (Maybe String) -> ClashFlag
+maybeStringFlag nm opt = ClashFlag
+  { cfFlags = \opts ->
+         strSuffixId nm opts (set opt . Just)
+      ++ unsetArg nm opts (set opt)
+  , cfRender = \opts ->
+      if isDefault opt opts
+      then []
+      else case opts ^. opt of
+        Nothing -> renderNoArg nm False
+        Just s -> renderSepSuffix nm s
+  }
+
+
+-- | Return three flags setting a 'Maybe' 'String'. If no argument is supplied, a
+-- default value is set. For example, if the default is @simpleDefault@, then:
+--
+--   * @-fclash-foo@ sets the option to @Just "simpleDefault"@
+--   * @-fclash-foo=wibble@ sets the option to @Just "wibble"@
+--   * @-fclash-no-foo@ sets the option to @Nothing@
+--
+maybeStringWithDefaultFlag ::
+  FlagName ->
+  -- | Default
+  String ->
+  Lens' ClashOpts (Maybe String) ->
+  ClashFlag
+maybeStringWithDefaultFlag flagName dflt opt = ClashFlag
+  { cfFlags = \opts ->
+         anySuffix flagName opts parser (set opt)
+      ++ unsetArg flagName opts (set opt)
+  , cfRender = \opts ->
+      if isDefault opt opts
+      then []
+      else case opts ^. opt of
+        Nothing -> renderNoArg flagName False
+        Just s | s == dflt -> renderSuffix flagName ""
+        Just s -> renderSuffix flagName s
+  }
+ where
+  parser :: String -> EwM IO (Maybe (Maybe String))
+  parser ['='] = pure (Just (Just dflt))
+  parser ('=':v) = pure (Just (Just v))
+  parser s = do
+    addErr ("Could not parse: " <> s)
+    pure Nothing
+
+maybeIntFlag :: FlagName -> Lens' ClashOpts (Maybe Word) -> ClashFlag
+maybeIntFlag = maybeReadableFlag
+
+maybeWordFlag :: FlagName -> Lens' ClashOpts (Maybe Word) -> ClashFlag
+maybeWordFlag = maybeReadableFlag
```

Comment:
I would just use `maybeReadableFlag` directly for these. Also `maybeIntFlag` has a `Maybe Word` which I assume should be `Maybe Int`

Reply:
> Also maybeIntFlag has a Maybe Word which I assume should be Maybe Int

Woops

> I would just use maybeReadableFlag directly for these.

This is actually wrong in another way: ints are passed as `-fclash-foo=3`, but `maybeReadableFlag` uses `-fclash-foo 3`.

Reply:
> This is actually wrong in another way

Then I'm not sure I understand why they're defined like `maybeIntFlag = maybeReadableFlag` and `maybeWordFlag = maybeReadableFlag`. Either you can just use `maybeReadableFlag` or these definitions are wrong, no?

Reply:
Yeah, the point of my comment was to point out my own code is wrong, not your comment. Sorry for the confusion.

> Either you can just use maybeReadableFlag or these definitions are wrong, no?

We've established a (frankly stupid) pattern that int/word is passed `-flike-this=3` while strings are passed `-flike-that str`. `maybeReadableFlag` interprets a string, so it will use the latter pattern. However, if we then use it to parse ints/words we're in the "wrong". So, it's not that `maybeReadableFlag` is wrong, but its usage for ints/words is.

I've put a comment somewhere that I think we should be liberal in what type of flag passing we accept for both.